### PR TITLE
Optional lambda function updated which periodically scans for old stacks

### PR DESCRIPTION
**Pull request by joleneyu**
Updated source code to add a scheduled Lambda function to periodically scan and delete the old stacks for the issue mentioned here https://github.com/lendi-au/StackJanitor/issues/58

This is a fallback mechanism that can remove stacks older than DEFAULT_EXPIRATION_PERIOD with the "stackjanitor: enabled" tag on the stack.

This optional scheduled Lambda function is handled by "./src/optional-handlers/searchDeleteCloudformationStacks.ts".

This optional function is not enabled by default. Update "serverless.yml" to enable this function.

### CODE REVIEW

File: serverless.yml [hunk 1]
```diff
@@ -148,6 +149,18 @@ functions:
           arn:
             Fn::GetAtt: [stackJanitorTable, StreamArn]
 
+  searchDeleteCloudformationStack:
+    handler: src/optional-handlers/searchDeleteCloudformationStack.handler
+    environment:
+      DEFAULT_EXPIRATION_PERIOD: ${self:custom.DEFAULT_EXPIRATION_PERIOD}
```
Comment (edify42) [inline serverless.yml:155]: Maybe have this as `${self:custom.DEFAULT_EXPIRATION_PERIOD} x 2` - not sure how math statements would resolve in yaml though

File: src/config.ts [hunk 2]
```diff
@@ -1,14 +1,24 @@
...
 const config = {
   DEFAULT_REGION: process.env.DEFAULT_REGION || "ap-southeast-2",
   DEFAULT_DYNAMODB_TABLE:
     process.env.DEFAULT_DYNAMODB_TABLE || "stackJanitorTable",
   DEFAULT_EXPIRATION_PERIOD:
     process.env.DEFAULT_EXPIRATION_PERIOD || SEVEN_DAYS,
   MAX_CLEANUP_RETRY: process.env.MAX_CLEANUP_RETRY || 5,
-  DELETE_INTERVAL
+  DELETE_INTERVAL,
+  DRY_RUN: isDryRun() || false
```
Comment (edify42) [inline src/config.ts:21]: no need for `|| false`

File: src/optional-handlers/searchDeleteCloudformationStack.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    "CREATE_FAILED",
+    "ROLLBACK_FAILED",
+    "ROLLBACK_COMPLETE",
+    "DELETE_FAILED",
+    "UPDATE_ROLLBACK_FAILED",
+    "UPDATE_ROLLBACK_COMPLETE",
+    "IMPORT_ROLLBACK_FAILED",
+    "IMPORT_ROLLBACK_COMPLETE"
+  ];
+  const newStacks1 = stacks.filter(stack => {
```
Comment (edify42) [inline src/optional-handlers/searchDeleteCloudformationStack.ts:20]: name this variable better - maybe `desiredStacks`

File: src/optional-handlers/searchDeleteCloudformationStack.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      ) {
+        return stack;
+      }
+    }
+  });
+  return newStacks2;
+}
+
+export function isStackExpired(stack: Stacks) {
+  const newStacks3 = stack.filter(stack => {
```
Comment (edify42) [inline src/optional-handlers/searchDeleteCloudformationStack.ts:42]: rename this variable to `expiredStacks`

File: src/optional-handlers/searchDeleteCloudformationStack.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  if (allStacks.Stacks) {
+    const stacks1 = returnStackStatus(allStacks.Stacks);
+    const stacks2 = returnStackTags(stacks1);
+    const stacks3 = isStackExpired(stacks2);
+    const stackNames = getStackName(stacks3);
+    stackNames.forEach(async (stackname: string) => {
+      await deleteStack(stackname);
+    });
+  }
+  return "No expired stacks found";
```
Comment (edify42) [inline src/optional-handlers/searchDeleteCloudformationStack.ts:90]: returned string will always say this which isn't true :)

File: src/optional-handlers/searchDeleteCloudformationStack.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  return cloudFormation.deleteStack({ StackName: stackName }).promise();
+}
+
+export const handler = async () => {
+  const cloudFormation = new AWS.CloudFormation();
+  AWS.config.update({ region: config.DEFAULT_REGION });
+  const allStacks = await cloudFormation.describeStacks().promise();
+  if (allStacks.Stacks) {
+    const stacks1 = returnStackStatus(allStacks.Stacks);
+    const stacks2 = returnStackTags(stacks1);
```
Comment (edify42) [inline src/optional-handlers/searchDeleteCloudformationStack.ts:83]: rename these variables in a similar meaningful way :)

**Comment by joleneyu** [comment]
Please see the new commit, it should resolve all the above you mentioned. Thank you:)

**Review by edify42 — APPROVED** [review]

File: src/optional-handlers/searchDeleteCloudformationStack.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+export function deleteStack(stackName: string) {
+  const cloudFormation = new AWS.CloudFormation();
+  console.log(`Deleting stack ${stackName}`);
+  return cloudFormation.deleteStack({ StackName: stackName }).promise();
+}
+
+export const handler = async () => {
+  const cloudFormation = new AWS.CloudFormation();
+  AWS.config.update({ region: config.DEFAULT_REGION });
+  const allStacks = await cloudFormation.describeStacks().promise();
```
[the remark this answers is not in the document]
Comment (joleneyu) [reply src/optional-handlers/searchDeleteCloudformationStack.ts]: Yes, paginated results can be returned. Updated describeAllStacks function with paginated result considered. Updated unit tests accordingly.
[https://github.com/lendi-au/StackJanitor/pull/60/files#diff-cb50987cff29e516744d787d660d8be0f0efaaaec6d53dfcb953cf2bee7407d7R9](url)

File: src/optional-handlers/searchDeleteCloudformationStack.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+export const handler = async () => {
+  const allStacks = await describeAllStacks();
+  if (allStacks) {
+    const desiredStacks = returnStackStatus(allStacks);
+    const stackjanitorEnabledStacks = returnStackTags(desiredStacks);
+    const expiredStacks = isStackExpired(stackjanitorEnabledStacks);
+    const stackNames = getStackName(expiredStacks);
+    if (stackNames.length === 0) {
+      return "No expired stacks found";
```
[the remark this answers is not in the document]
Comment (joleneyu) [reply src/optional-handlers/searchDeleteCloudformationStack.ts]: I verified that the Lambda function sent back the 200/202 code for a successful request. For the Request-Response invocation type, this status code is 200. For the Event invocation type, this status code is 202.

File: src/optional-handlers/searchDeleteCloudformationStack.test.ts [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+  });
+
+  test("should test stub is working", async () => {
+    const result: Stacks = [];
+    const cfnMock = sinon.stub();
+    const mockResolves: DescribeStacksOutput = {};
+    AWSMock.mock("CloudFormation", "describeStacks", cfnMock);
+    cfnMock.resolves(mockResolves);
+    const mystacks = await describeAllStacks();
+    console.log(cfnMock.called);
```
[the remark this answers is not in the document]
Comment (joleneyu) [reply src/optional-handlers/searchDeleteCloudformationStack.test.ts]: Removed!

**Review by luizfelipepg — APPROVED** [review]
Thanks for checking it :)

**Pull request merged**